File: src/main/java/com/github/Doomsdayrs/Jikan4java/core/search/animeManga/AnimeSearch.java
package com.github.Doomsdayrs.Jikan4java.core.search.animeManga;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.Doomsdayrs.Jikan4java.enums.SortBy;
import com.github.Doomsdayrs.Jikan4java.enums.genres.AnimeGenres;
import com.github.Doomsdayrs.Jikan4java.enums.search.Types;
import com.github.Doomsdayrs.Jikan4java.enums.search.animeManga.Ratings;
import com.github.Doomsdayrs.Jikan4java.enums.search.animeManga.orderBy.AnimeOrderBy;
import com.github.Doomsdayrs.Jikan4java.enums.search.animeManga.subType.SubTypes;
import com.github.Doomsdayrs.Jikan4java.enums.status.AnimeStati;
import com.github.Doomsdayrs.Jikan4java.types.Main.Anime.Anime;
import com.github.Doomsdayrs.Jikan4java.types.Main.Anime.AnimePage.AnimePage;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import org.json.simple.parser.JSONParser;

import java.util.concurrent.CompletableFuture;

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 * ====================================================================
 * Jikan4java
 * 13 / 05 / 2019
 *
 * @author github.com/doomsdayrs
 */
public class AnimeSearch extends AnimeMangaSearch<AnimePage> {

    public AnimeSearch() {
        super(Types.ANIME);
    }

    public AnimeSearch(ObjectMapper mapper) {
        super(Types.ANIME, mapper);
    }

    public AnimeSearch(Types type, OkHttpClient client) {
        super(type, client);
    }

    public AnimeSearch(Types type, JSONParser jsonParser) {
        super(type, jsonParser);
    }

    public AnimeSearch(Types type, Request.Builder builder) {
        super(type, builder);
    }

    @Override
    public CompletableFuture<Anime> getFirst() {
        return super.getFirst();
    }

    @Override
    public AnimeSearch setQuery(String title) {
        super.setQuery(title);
        return this;
    }

    @Override
    public AnimeSearch setPage(int page) {
        super.setPage(page);
        return this;
    }

    public AnimeSearch setStatus(AnimeStati status) {
        super.setStatus(status);
        return this;
    }

    @Override
    public AnimeSearch setRating(Ratings rating) {
        super.setRating(rating);
        return this;
    }

    @Override
    public AnimeSearch setScore(float score) {
        super.setScore(score);
        return this;
    }

    @Override
    public AnimeSearch toggleGenres() {
        super.toggleGenres();
        return this;
    }

    @Override
    public AnimeSearch setLimit(int pages) {
        super.setLimit(pages);
        return this;
    }

    public AnimeSearch addGenre(AnimeGenres animeGenres) {
        super.addGenre(animeGenres);
        return this;
    }

    @Override
    public AnimeSearch setStartDate(int yyyy, int mm, int dd) {
        super.setStartDate(yyyy, mm, dd);
        return this;
    }

    @Override
    public AnimeSearch setEndDate(int yyyy, int mm, int dd) {
        super.setEndDate(yyyy, mm, dd);
        return this;
    }


    public AnimeSearch orderBy(AnimeOrderBy orderBy) {
        super.orderBy(orderBy);
        return this;
    }

    @Override
    public AnimeSearch sortBy(SortBy sortBy) {
        super.sortBy(sortBy);
        return this;
    }

    @Override
    public AnimeSearch setType(SubTypes subTypes) {
        super.setType(subTypes);
        return this;
    }

    @Override
    public CompletableFuture<Anime> getByID(int id) {
        return super.getByID(id);
    }
}

File: src/commonMain/kotlin/com/github/doomsdayrs/jikan4java/core/search/animemanga/AnimeSearch.kt
package com.github.doomsdayrs.jikan4java.core.search.animemanga

import com.github.doomsdayrs.jikan4java.data.enums.genres.AnimeGenres
import com.github.doomsdayrs.jikan4java.data.enums.search.Types
import com.github.doomsdayrs.jikan4java.data.enums.search.animemanga.orderby.AnimeOrderBy
import com.github.doomsdayrs.jikan4java.data.enums.status.AnimeStati
import com.github.doomsdayrs.jikan4java.data.model.main.genresearch.anime.GenreSearchAnime
import com.github.doomsdayrs.jikan4java.data.model.main.genresearch.anime.GenreSearchAnimePage

/*
 * This file is part of Jikan4java.
 *
 * Jikan4java is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Jikan4java is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Jikan4java.  If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * Jikan4java
 * 13 / 05 / 2019
 *
 * @author github.com/doomsdayrs
 */
class AnimeSearch :
	AnimeMangaSearch< AnimeSearch, AnimeStati, AnimeGenres, AnimeOrderBy, GenreSearchAnime, GenreSearchAnimePage>(
		Types.ANIME,
	)